ethcore/src/client.rs

```diff
@@ -243,85 +243,117 @@ impl Client {
+				mark_block_as_bad();
+				break;
 			};
-			// build last hashes
-			let mut last_hashes = LastHashes::new();
-			last_hashes.resize(256, H256::new());
-			last_hashes[0] = header.parent_hash.clone();
-			for i in 0..255 {
-				match self.chain.read().unwrap().block_details(&last_hashes[i]) {
-					Some(details) => {
-						last_hashes[i + 1] = details.parent.clone();
-					},
-					None => break,
-				}
-			}
 
+			// Enact Verified Block
+			let parent = chain_has_parent.unwrap();
+			let last_hashes = self.build_last_hashes(header);
 			let db = self.state_db.lock().unwrap().clone();
-			let result = match enact_verified(&block, self.engine.deref().deref(), db, &parent, &last_hashes) {
-				Ok(b) => b,
-				Err(e) => {
-					warn!(target: "client", "Block import failed for #{} ({})\nError: {:?}", header.number(), header.hash(), e);
-					bad.insert(block.header.hash());
-					self.block_queue.write().unwrap().mark_as_bad(&header.hash());
-					break;
-				}
+
+			let enact_result = enact_verified(&block, engine, db, &parent, &last_hashes);
+			if let Err(e) = enact_result {
+				warn!(target: "client", "Block import failed for #{} ({})\nError: {:?}", header.number(), header.hash(), e);
+				mark_block_as_bad();
+				break;
 			};
-			if let Err(e) = verify_block_final(&header, result.block().header()) {
+
+			// Final Verification
+			let enact_result = enact_result.unwrap();
+			if let Err(e) = verify_block_final(&header, enact_result.block().header()) {
 				warn!(target: "client", "Stage 4 block verification failed for #{} ({})\nError: {:?}", header.number(), header.hash(), e);
-				self.block_queue.write().unwrap().mark_as_bad(&header.hash());
+				mark_block_as_bad();
 				break;
 			}
 
-			good_blocks.push(header.hash().clone());
-
+			// Insert block
 			self.chain.write().unwrap().insert_block(&block.bytes); //TODO: err here?
-			let ancient = if header.number() >= HISTORY { Some(header.number() - HISTORY) } else { None };
-			match result.drain().commit(header.number(), &header.hash(), ancient.map(|n|(n, self.chain.read().unwrap().block_hash(n).unwrap()))) {
-				Ok(_) => (),
-				Err(e) => {
-					warn!(target: "client", "State DB commit failed: {:?}", e);
-					break;
-				}
+			good_blocks.push(header.hash());
+
+			let ancient = if header.number() >= HISTORY {
+				let n = header.number() - HISTORY;
+				let chain = self.chain.read().unwrap();
+				Some((n, chain.block_hash(n).unwrap()))
+			} else {
+				None
+			};
+
+			// Commit results
+			let commit_result = enact_result.drain().commit(header.number(), &header.hash(), ancient);
+			if let Err(e) = commit_result {
+				warn!(target: "client", "State DB commit failed: {:?}", e);
+				break;
 			}
+
 			self.report.write().unwrap().accrue_block(&block);
 			trace!(target: "client", "Imported #{} ({})", header.number(), header.hash());
-			ret += 1;
+			imported += 1;
```

Comment:
We track this only because state DB commit may fail (in that case `imported == good_blocks.len() - 1` in every other imported == good_blocks.len()`) - could I refactor this further to avoid this variable at all?